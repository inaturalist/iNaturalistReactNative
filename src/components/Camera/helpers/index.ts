import { Alert } from "react-native";
import { CameraRuntimeError } from "react-native-vision-camera";

import { log } from "../../../../react-native-logs.config";

const logger = log.extend( "CameraView" );

type Event = {
  log: string;
}

const handleClassifierError = ( error: CameraRuntimeError ) => {
  // When we hit this error, there is an error with the classifier.
  logger.error( "handleClassifierError: ", JSON.stringify( error ) );
  Alert.alert( "error", JSON.stringify( error ) );
};

const handleDeviceNotSupported = ( error: CameraRuntimeError ) => {
  // When we hit this error, something with the current device is not supported.
  Alert.alert( "error", JSON.stringify( error ) );
};

const handleCaptureError = ( error: CameraRuntimeError ) => {
  // When we hit this error, taking a photo did not work correctly
  logger.error( "handleCaptureError: ", JSON.stringify( error ) );
  Alert.alert( "error", JSON.stringify( error ) );
};

const handleCameraError = ( error: CameraRuntimeError ) => {
  // This error is thrown when it does not fit in any of the above categories.
  logger.error( "handleCameraError ", JSON.stringify( error ) );
  Alert.alert( "error", JSON.stringify( error ) );
};

const handleLog = ( event: Event ) => {
  logger.debug( event.log );
};

export {
  handleCameraError,
  handleCaptureError,
  handleClassifierError,
  handleDeviceNotSupported,
  handleLog
};
