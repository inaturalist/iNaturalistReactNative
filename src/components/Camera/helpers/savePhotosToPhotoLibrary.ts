import { CameraRoll } from "@react-native-camera-roll/camera-roll";
import {
  permissionResultFromMultiple,
  READ_WRITE_MEDIA_PERMISSIONS
} from "components/SharedComponents/PermissionGateContainer.tsx";
import { t } from "i18next";
import {
  Alert
} from "react-native";
import { checkMultiple, RESULTS } from "react-native-permissions";
import { log } from "sharedHelpers/logger";

import { displayName as appName } from "../../../../app.json";

const logger = log.extend( "savePhotosToPhotoLibrary" );

// Save URIs to camera photo library (if a photo was taken using the app,
// we want it accessible in the camera's folder, as if the user has taken those photos
// via their own camera app).
// One could argue this is a private method and shouldn't be exported and
// doesn't need to be tested... but hooks are complicated and this hook might
// be too complicated, so this at least makes it easy to test this one part
// ~~~kueda20240614
// $FlowIgnore
export async function savePhotosToPhotoLibrary(
  uris: [string],
  location: Object
) {
  const readWritePermissionResult = permissionResultFromMultiple(
    await checkMultiple( READ_WRITE_MEDIA_PERMISSIONS )
  );
  const savedPhotoUris = await uris.reduce(
    async ( memo, uri ) => {
      const savedUris = await memo;
      try {
        const saveOptions = {};
        // One quirk of CameraRoll is that if you want to write to an album, you
        // need readwrite permission, but we don't want to ask for that here
        // b/c it might come immediately after asking for *add only*
        // permission, so we're checking to see if we have that permission
        // and skipping the album if we don't
        if ( readWritePermissionResult === RESULTS.GRANTED ) {
          saveOptions.type = "photo";
          // Note: we do not translate our brand name, so this should not be
          // globalized
          saveOptions.album = appName;
        }
        if ( location ) {
          saveOptions.latitude = location.latitude;
          saveOptions.longitude = location.longitude;
          saveOptions.horizontalAccuracy = location.positional_accuracy;
        }
        const savedPhotoUri = await CameraRoll.save( uri, saveOptions );
        savedUris.push( savedPhotoUri );
        return savedUris;
      } catch ( cameraRollSaveError ) {
        // should never get here since in usePrepareStoreAndNavigate we check for device full
        // and skip saving to photo library
        if (
          cameraRollSaveError.message.match( /No space left on device/ )
          || cameraRollSaveError.message.match( /PHPhotosErrorDomain error 3305/ )
        ) {
          Alert.alert(
            t( "Not-enough-space-left-on-device" ),
            t( "Not-enough-space-left-on-device-try-again" ),
            [{ text: t( "OK" ) }]
          );
          return savedUris;
        }
        // This means an iOS user denied access
        // (https://developer.apple.com/documentation/photokit/phphotoserror/code/accessuserdenied).
        // In theory we should not even have called this function when that
        // happens, but we're still seeing this in the logs. They should be
        // prompted to grant permission the next time they try so this is
        // probably safe to ignore.
        if ( !cameraRollSaveError.message.match( /error 3311/ ) ) {
          logger.error( cameraRollSaveError );
          return savedUris;
        }
        throw cameraRollSaveError;
      }
    },
    // We need the initial value even if we're not using it, otherwise reduce
    // will treat the first item in the array as the initial value and not
    // call the reducer function on it
    Promise.resolve( [] )
  );
  return savedPhotoUris;
}

export default savePhotosToPhotoLibrary;
