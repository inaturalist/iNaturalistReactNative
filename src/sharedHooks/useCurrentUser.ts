import { RealmContext } from "providers/contexts";
import { useEffect, useState } from "react";
import Realm from "realm";
import type { RealmUser } from "realmModels/types";

const { useRealm } = RealmContext;

const useCurrentUser = ( ): RealmUser | null => {
  const realm = useRealm( );
  const [currentUser, setCurrentUser] = useState<RealmUser | null>( null );
  useEffect( ( ) => {
    const realmResults = realm.objects( "User" ).filtered( "signedIn == true" );

    // Sets the current user if one is detected, unsets it if not
    const listener = (
      collection: [Realm.Object<RealmUser>],
      changes: Realm.CollectionChangeSet
    ) => {
      if (
        ( changes.deletions && changes.deletions.length > 0 )
        || !collection[0]?.isValid( )
      ) {
        setCurrentUser( null );
      } else {
        setCurrentUser( collection[0] );
      }
    };

    // Run the listener on the results for the first time
    listener( realmResults, {} );

    // Add the listener
    realmResults?.addListener( listener );

    // Remove the listener when this dismounts
    return ( ) => {
      realmResults?.removeListener( listener );
    };
  }, [realm] );

  return currentUser?.isValid( )
    ? currentUser
    : null;
};

export default useCurrentUser;
