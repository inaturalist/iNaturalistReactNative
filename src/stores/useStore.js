import _ from "lodash";
import { MMKV } from "react-native-mmkv";
import { create } from "zustand";
import { createJSONStorage, persist } from "zustand/middleware";

import createExploreSlice from "./createExploreSlice";
import createLayoutSlice from "./createLayoutSlice";
import createMyObsSlice from "./createMyObsSlice";
import createObservationFlowSlice from "./createObservationFlowSlice";
import createRootExploreSlice from "./createRootExploreSlice";
import createSyncObservationsSlice from "./createSyncObservationsSlice";
import createUploadObservationsSlice from "./createUploadObservationsSlice";

export const storage = new MMKV( );

// TODO do *not* export this. This allows any consumer to overwrite *any* part
// of state, circumventing any getter/setter logic we have in the stores. If
// you need to modify state, you should be doing so through a store.
export const zustandStorage = {
  setItem: ( name, value ) => storage.set( name, value ),
  getItem: name => {
    const value = storage.getString( name ) || storage.getNumber( name );
    return value ?? null;
  },
  removeItem: name => storage.delete( name )
};

// Using slices to separate store for Explore and Observation creation flow
// https://docs.pmnd.rs/zustand/guides/slices-pattern
const useStore = create( persist(
  ( ...args ) => {
    // Let's make our slices
    const slices = [
      createExploreSlice( ...args ),
      createLayoutSlice( ...args ),
      createMyObsSlice( ...args ),
      createObservationFlowSlice( ...args ),
      createRootExploreSlice( ...args ),
      createSyncObservationsSlice( ...args ),
      createUploadObservationsSlice( ...args )
    ];

    // Now let's make sure they're not clobbering each other because
    // everything in Zustand state exists in a single namespace and
    // clobbering can happen silently... which is bad.
    const allKeys = slices.map( slice => Object.keys( slice ) ).flat( );
    const keyCounts = allKeys.reduce(
      ( memo, curr ) => {
        memo[curr] ||= 0;
        memo[curr] += 1;
        return memo;
      },
      {}
    );
    const nonUniqueKeys = Object.keys( keyCounts ).reduce(
      ( memo, curr ) => {
        if ( keyCounts[curr] > 1 ) memo.push( curr );
        return memo;
      },
      []
    );
    if ( nonUniqueKeys.length > 0 ) {
      throw new Error(
        `You have multiple Zustand slices with the following keys: ${nonUniqueKeys}`
      );
    }

    // All good? Now let's combine them into one enormous hideous object
    return slices.reduce(
      ( memo, curr ) => ( { ...memo, ...curr } ),
      {}
    );
  },
  {
    name: "persisted-zustand",
    partialize: state => ( {
      // Vestigial un-namespaced values in the layout slice
      isAdvancedUser: state.isAdvancedUser,
      obsDetailsTab: state.obsDetailsTab,

      // Dynamically select all values in the layout slice's namespace
      layout: ( Object.keys( state.layout ).reduce( ( memo, key ) => {
        if ( typeof ( state.layout[key] ) !== "function" ) {
          memo[key] = state.layout[key];
        }
        return memo;
      }, {} ) )
    } ),
    storage: createJSONStorage( () => zustandStorage ),
    // We need to deep merge to persist nested objects, like layout
    // https://zustand.docs.pmnd.rs/middlewares/persist#persisting-a-state-with-nested-objects
    merge: ( persisted, current ) => _.merge( current, persisted )
  }
) );

export default useStore;
