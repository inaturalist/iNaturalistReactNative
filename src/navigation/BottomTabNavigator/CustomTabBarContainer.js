// @flow
import { useDrawerStatus } from "@react-navigation/drawer";
import { getCurrentRoute, navigateToTabStack } from "navigation/navigationUtils.ts";
import {
  SCREEN_NAME_NOTIFICATIONS,
  SCREEN_NAME_OBS_LIST,
  SCREEN_NAME_ROOT_EXPLORE
} from "navigation/StackNavigators/TabStackNavigator";
import type { Node } from "react";
import React, { useMemo, useState } from "react";
import User from "realmModels/User.ts";
import { useCurrentUser, useTranslation } from "sharedHooks";

import CustomTabBar from "./CustomTabBar";

const DRAWER_ID = "OPEN_DRAWER";

type Props = {
  navigation: Object
};

const tabIDByRoute = {
  [SCREEN_NAME_NOTIFICATIONS]: SCREEN_NAME_NOTIFICATIONS,
  [SCREEN_NAME_OBS_LIST]: SCREEN_NAME_OBS_LIST,
  [SCREEN_NAME_ROOT_EXPLORE]: SCREEN_NAME_ROOT_EXPLORE
};

const CustomTabBarContainer = ( { navigation }: Props ): Node => {
  const { t } = useTranslation( );
  const currentUser = useCurrentUser( );
  const [activeTab, setActiveTab] = useState( SCREEN_NAME_OBS_LIST );
  const isDrawerOpen = useDrawerStatus() === "open";
  const route = getCurrentRoute();
  const currentRoute = route?.name || "";
  const currentActiveTab = tabIDByRoute[currentRoute] || activeTab;

  // Update activeTab only if it has changed to a different tab
  if ( currentActiveTab !== activeTab ) {
    setActiveTab( currentActiveTab );
  }

  const tabs = useMemo( ( ) => ( [
    {
      icon: "hamburger-menu",
      testID: DRAWER_ID,
      accessibilityLabel: t( "Menu" ),
      accessibilityHint: t( "Opens-the-side-drawer-menu" ),
      size: 32,
      onPress: ( ) => {
        navigation.openDrawer( );
      },
      active: isDrawerOpen
    },
    {
      icon: "magnifying-glass",
      testID: SCREEN_NAME_ROOT_EXPLORE,
      accessibilityLabel: t( "Explore" ),
      accessibilityHint: t( "Navigates-to-explore" ),
      size: 31,
      onPress: ( ) => {
        navigateToTabStack( "RootExplore" );
      },
      active: SCREEN_NAME_ROOT_EXPLORE === activeTab
    },
    {
      icon: "person",
      userIconUri: User.uri( currentUser ),
      testID: "NavButton.personIcon",
      accessibilityLabel: t( "My-Observations--bottom-tab" ),
      accessibilityHint: t( "Navigates-to-your-observations" ),
      size: 40,
      onPress: ( ) => {
        navigateToTabStack( "ObsList" );
      },
      active: SCREEN_NAME_OBS_LIST === activeTab
    },
    {
      icon: "notifications-bell",
      testID: SCREEN_NAME_NOTIFICATIONS,
      accessibilityLabel: t( "Notifications--bottom-tab" ),
      accessibilityHint: t( "Navigates-to-notifications" ),
      size: 32,
      onPress: ( ) => {
        navigateToTabStack( "Notifications" );
      },
      active: SCREEN_NAME_NOTIFICATIONS === activeTab
    }
  ] ), [
    activeTab,
    currentUser,
    isDrawerOpen,
    navigation,
    t
  ] );

  return (
    <CustomTabBar
      tabs={tabs}
    />
  );
};

export default CustomTabBarContainer;
