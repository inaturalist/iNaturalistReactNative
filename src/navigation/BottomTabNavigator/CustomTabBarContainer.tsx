import type { BottomTabBarProps } from "@react-navigation/bottom-tabs";
import { useDrawerStatus } from "@react-navigation/drawer";
import {
  SCREEN_NAME_NOTIFICATIONS,
  SCREEN_NAME_OBS_LIST,
  SCREEN_NAME_ROOT_EXPLORE
} from "navigation/StackNavigators/TabStackNavigator";
import React, { useMemo } from "react";
import User from "realmModels/User.ts";
import { useCurrentUser, useTranslation } from "sharedHooks";

import CustomTabBar from "./CustomTabBar";

const DRAWER_ID = "OPEN_DRAWER";

interface TabConfig {
  icon: string;
  testID: string;
  accessibilityLabel: string;
  accessibilityHint: string;
  size: number;
  onPress: () => void;
  active: boolean;
  userIconUri?: string;
}

type TabName = "ObservationsTab" | "ExploreTab" | "NotificationsTab";

type ScreenName =
  | typeof SCREEN_NAME_OBS_LIST
  | typeof SCREEN_NAME_ROOT_EXPLORE
  | typeof SCREEN_NAME_NOTIFICATIONS;

type Props = BottomTabBarProps;

const CustomTabBarContainer: React.FC<Props> = ( { navigation, state } ) => {
  const { t } = useTranslation( );
  const currentUser = useCurrentUser( );
  const isDrawerOpen = useDrawerStatus() === "open";

  const activeTabIndex = state?.index;
  const activeTabName = state?.routes[activeTabIndex]?.name as TabName;

  const userIconUri = useMemo( ( ) => User.uri( currentUser ), [currentUser] );

  const getActiveTab = ( ): ScreenName => {
    switch ( activeTabName ) {
      case "ObservationsTab": return SCREEN_NAME_OBS_LIST;
      case "ExploreTab": return SCREEN_NAME_ROOT_EXPLORE;
      case "NotificationsTab": return SCREEN_NAME_NOTIFICATIONS;
      default: return SCREEN_NAME_OBS_LIST;
    }
  };

  const activeTab = getActiveTab( );

  const tabs: TabConfig[] = useMemo( ( ) => ( [
    {
      icon: "hamburger-menu",
      testID: DRAWER_ID,
      accessibilityLabel: t( "Menu" ),
      accessibilityHint: t( "Opens-the-side-drawer-menu" ),
      size: 32,
      onPress: ( ) => {
        navigation.openDrawer( );
      },
      active: isDrawerOpen
    },
    {
      icon: "magnifying-glass",
      testID: SCREEN_NAME_ROOT_EXPLORE,
      accessibilityLabel: t( "Explore" ),
      accessibilityHint: t( "Navigates-to-explore" ),
      size: 31,
      onPress: ( ) => {
        navigation.navigate( "ExploreTab", {
          screen: "RootExplore"
        } );
      },
      active: SCREEN_NAME_ROOT_EXPLORE === activeTab
    },
    {
      icon: "person",
      userIconUri,
      testID: "NavButton.personIcon",
      accessibilityLabel: t( "My-Observations--bottom-tab" ),
      accessibilityHint: t( "Navigates-to-your-observations" ),
      size: 40,
      onPress: ( ) => {
        navigation.navigate( "ObservationsTab", {
          screen: "ObsList"
        } );
      },
      active: SCREEN_NAME_OBS_LIST === activeTab
    },
    {
      icon: "notifications-bell",
      testID: SCREEN_NAME_NOTIFICATIONS,
      accessibilityLabel: t( "Notifications--bottom-tab" ),
      accessibilityHint: t( "Navigates-to-notifications" ),
      size: 32,
      onPress: ( ) => {
        navigation.navigate( "NotificationsTab", {
          screen: "Notifications"
        } );
      },
      active: SCREEN_NAME_NOTIFICATIONS === activeTab
    }
  ] ), [
    activeTab,
    userIconUri,
    isDrawerOpen,
    navigation,
    t
  ] );

  return (
    <CustomTabBar
      tabs={tabs}
    />
  );
};

export default CustomTabBarContainer;
