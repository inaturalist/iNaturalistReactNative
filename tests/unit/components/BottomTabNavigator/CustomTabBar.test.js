import {
  useNetInfo
} from "@react-native-community/netinfo";
import { screen } from "@testing-library/react-native";
import CustomTabBarContainer from "navigation/BottomTabNavigator/CustomTabBarContainer.tsx";
import React from "react";
import * as useCurrentUser from "sharedHooks/useCurrentUser.ts";
import useStore from "stores/useStore";
import factory from "tests/factory";
import faker from "tests/helpers/faker";
import { renderComponent } from "tests/helpers/render";

jest.mock( "@react-navigation/drawer", ( ) => {
  const actualNav = jest.requireActual( "@react-navigation/drawer" );
  return {
    ...actualNav,
    useDrawerStatus: jest.fn( ( ) => false )
  };
} );

const initialPersistedStoreState = useStore.getState( );

const mockUser = factory( "LocalUser", {
  login: faker.internet.userName( ),
  icon_url: faker.image.url( )
} );

jest.mock( "sharedHooks/useCurrentUser", ( ) => ( {
  __esModule: true,
  default: () => undefined
} ) );

jest.mock( "sharedHooks/useAuthenticatedQuery", () => ( {
  __esModule: true,
  default: () => ( {
    data: 0
  } )
} ) );

describe( "CustomTabBar", () => {
  beforeEach( ( ) => {
    jest.useFakeTimers();
  } );

  // it( "should render correctly", async () => {
  //   renderComponent( <CustomTabBarContainer navigation={jest.fn( )} /> );

  //   await expect( screen ).toMatchSnapshot();
  // } );

  it( "should not have accessibility errors", async () => {
    const tabBar = <CustomTabBarContainer navigation={jest.fn( )} />;

    await expect( tabBar ).toBeAccessible();
  } );

  it( "should display person icon while user is logged out", async () => {
    renderComponent( <CustomTabBarContainer navigation={jest.fn( )} /> );

    const personIcon = screen.getByTestId( "NavButton.personIcon" );
    await expect( personIcon ).toBeVisible( );
  } );

  it( "should display avatar while user is logged in", async () => {
    jest.spyOn( useCurrentUser, "default" ).mockImplementation( () => mockUser );
    renderComponent( <CustomTabBarContainer navigation={jest.fn( )} /> );

    const avatar = screen.getByTestId( "UserIcon.photo" );
    await expect( avatar ).toBeVisible( );
  } );

  it( "should display person icon when connectivity is low", async ( ) => {
    useNetInfo.mockImplementation( ( ) => ( { isConnected: false } ) );
    renderComponent( <CustomTabBarContainer navigation={jest.fn( )} /> );

    const personIcon = screen.getByTestId( "NavButton.personIcon" );
    await expect( personIcon ).toBeVisible( );
  } );
} );

describe( "CustomTabBar with advanced user layout", () => {
  beforeAll( ( ) => {
    useStore.setState( {
      layout: {
        isAllAddObsOptionsMode: true
      }
    } );
  } );

  afterAll( ( ) => {
    useStore.setState( initialPersistedStoreState );
  } );

  beforeEach( ( ) => {
    jest.resetAllMocks();
    useNetInfo.mockImplementation( ( ) => ( { isConnected: true } ) );
    // Re-establish the safe area context mock after resetAllMocks
    const safeAreaContext = require( "react-native-safe-area-context" );
    safeAreaContext.useSafeAreaInsets.mockImplementation( () => ( {
      top: 0,
      right: 0,
      bottom: 0,
      left: 0
    } ) );
  } );

  it( "should render correctly", async () => {
    renderComponent( <CustomTabBarContainer navigation={jest.fn( )} /> );

    await expect( screen ).toMatchSnapshot();
  } );

  it( "should not have accessibility errors", async () => {
    const tabBar = <CustomTabBarContainer navigation={jest.fn( )} />;

    await expect( tabBar ).toBeAccessible();
  } );
} );
