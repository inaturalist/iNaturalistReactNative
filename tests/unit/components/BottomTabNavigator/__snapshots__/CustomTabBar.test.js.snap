// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomTabBar with advanced user layout should render correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    className="flex-1"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      accessibilityRole="tablist"
      style={
        [
          [
            {
              "flexDirection": "row",
            },
            {
              "backgroundColor": "#ffffff",
            },
            {
              "justifyContent": "space-evenly",
            },
            {
              "paddingBottom": 4,
              "paddingLeft": 4,
              "paddingRight": 4,
              "paddingTop": 4,
            },
            {
              "marginBottom": 0,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 0,
            },
            {
              "elevation": 20,
              "shadowColor": "rgba( 0, 0, 0, 0.25 )",
              "shadowOffset": {
                "height": -2,
                "width": 0,
              },
              "shadowOpacity": 0.25,
              "shadowRadius": 2,
            },
          ],
        ]
      }
      testID="CustomTabBar"
    >
      <View
        accessibilityHint="Opens the side drawer menu."
        accessibilityLabel="Menu"
        accessibilityRole="tab"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": false,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            [
              {
                "alignItems": "center",
              },
              {
                "width": "20%",
              },
              {
                "justifyContent": "flex-end",
              },
            ],
          ]
        }
      >
        <View
          style={
            [
              [
                {
                  "height": 44,
                  "width": 44,
                },
                {
                  "alignItems": "center",
                  "justifyContent": "center",
                },
                false,
                undefined,
              ],
            ]
          }
          testID="OPEN_DRAWER"
        >
          <View
            style={
              [
                [
                  {
                    "position": "relative",
                  },
                  {
                    "left": 0.2,
                  },
                  {
                    "top": 0.1,
                  },
                ],
              ]
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#454545",
                    "fontSize": 32,
                  },
                  null,
                  {
                    "fontFamily": "INatIcon",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
        <Text
          maxFontSizeMultiplier={1.2}
          numberOfLines={1}
          style={
            [
              {
                "color": "#454545",
              },
              {
                "fontFamily": "Lato-Regular",
              },
              [
                {
                  "textAlign": "left",
                },
                [
                  {
                    "fontSize": 11,
                    "lineHeight": 13.2,
                  },
                  {
                    "fontFamily": "Lato-Regular",
                  },
                  [
                    {
                      "color": "#454545",
                    },
                  ],
                ],
              ],
            ]
          }
        >
          Menu
        </Text>
      </View>
      <View
        accessibilityHint="Navigates to explore"
        accessibilityLabel="Explore"
        accessibilityRole="tab"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": false,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            [
              {
                "alignItems": "center",
              },
              {
                "width": "20%",
              },
              {
                "justifyContent": "flex-end",
              },
            ],
          ]
        }
      >
        <View
          style={
            [
              [
                {
                  "height": 44,
                  "width": 44,
                },
                {
                  "alignItems": "center",
                  "justifyContent": "center",
                },
                false,
                undefined,
              ],
            ]
          }
          testID="RootExplore"
        >
          <View
            style={
              [
                [
                  {
                    "position": "relative",
                  },
                  {
                    "left": 0.2,
                  },
                  {
                    "top": 0.1,
                  },
                ],
              ]
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#454545",
                    "fontSize": 40,
                  },
                  null,
                  {
                    "fontFamily": "INatIcon",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
        <Text
          maxFontSizeMultiplier={1.2}
          numberOfLines={1}
          style={
            [
              {
                "color": "#454545",
              },
              {
                "fontFamily": "Lato-Regular",
              },
              [
                {
                  "textAlign": "left",
                },
                [
                  {
                    "fontSize": 11,
                    "lineHeight": 13.2,
                  },
                  {
                    "fontFamily": "Lato-Regular",
                  },
                  [
                    {
                      "color": "#454545",
                    },
                  ],
                ],
              ],
            ]
          }
        >
          Explore
        </Text>
      </View>
      <View
        style={
          [
            [
              {
                "width": 69,
              },
              {
                "height": 60,
              },
              {
                "marginLeft": 12,
                "marginRight": 12,
              },
            ],
          ]
        }
      >
        <View
          style={
            [
              [
                {
                  "position": "absolute",
                },
                {
                  "top": -13,
                },
              ],
            ]
          }
        >
          <View
            accessibilityHint="Shows observation creation options"
            accessibilityLabel="Add observations"
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                [
                  {
                    "width": 69,
                  },
                  {
                    "height": 69,
                  },
                  {
                    "marginBottom": 5,
                  },
                  {
                    "borderBottomLeftRadius": 9999,
                    "borderBottomRightRadius": 9999,
                    "borderTopLeftRadius": 9999,
                    "borderTopRightRadius": 9999,
                  },
                  {
                    "overflow": "hidden",
                  },
                  {
                    "elevation": 5,
                    "shadowColor": "rgba( 0, 0, 0, 0.25 )",
                    "shadowOffset": {
                      "height": 2,
                      "width": 0,
                    },
                    "shadowOpacity": 0.25,
                    "shadowRadius": 2,
                  },
                ],
              ]
            }
            testID="add-obs-button"
          >
            <BVLinearGradient
              angle={156.95}
              colors={
                [
                  4285836288,
                  4280909703,
                ]
              }
              endPoint={
                {
                  "x": 0.5,
                  "y": 1,
                }
              }
              locations={null}
              startPoint={
                {
                  "x": 0.5,
                  "y": 0,
                }
              }
              useAngle={true}
            >
              <View
                style={
                  [
                    [
                      {
                        "flexGrow": 1,
                      },
                      {
                        "aspectRatio": 1,
                      },
                      {
                        "display": "flex",
                      },
                      {
                        "alignItems": "center",
                      },
                      {
                        "justifyContent": "center",
                      },
                    ],
                  ]
                }
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    [
                      {
                        "color": "#ffffff",
                        "fontSize": 31,
                      },
                      null,
                      {
                        "fontFamily": "INatIcon",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      {},
                    ]
                  }
                >
                  
                </Text>
              </View>
            </BVLinearGradient>
          </View>
        </View>
      </View>
      <View
        accessibilityHint="Navigates to your observations"
        accessibilityLabel="Me"
        accessibilityRole="tab"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": true,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            [
              {
                "alignItems": "center",
              },
              {
                "width": "20%",
              },
              {
                "justifyContent": "flex-end",
              },
            ],
          ]
        }
      >
        <View
          style={
            [
              [
                {
                  "height": 44,
                  "width": 44,
                },
                {
                  "alignItems": "center",
                  "justifyContent": "center",
                },
                false,
                undefined,
              ],
            ]
          }
          testID="NavButton.personIcon"
        >
          <View
            style={
              [
                [
                  {
                    "position": "relative",
                  },
                  {
                    "left": 0.2,
                  },
                  {
                    "top": 0.1,
                  },
                ],
              ]
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#74AC00",
                    "fontSize": 40,
                  },
                  null,
                  {
                    "fontFamily": "INatIcon",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
        <Text
          maxFontSizeMultiplier={1.2}
          numberOfLines={1}
          style={
            [
              {
                "color": "#454545",
              },
              {
                "fontFamily": "Lato-Regular",
              },
              [
                {
                  "textAlign": "left",
                },
                [
                  {
                    "fontSize": 11,
                    "lineHeight": 13.2,
                  },
                  {
                    "fontFamily": "Lato-Regular",
                  },
                  [
                    {
                      "color": "#74AC00",
                    },
                  ],
                ],
              ],
            ]
          }
        >
          Me
        </Text>
      </View>
      <View
        accessibilityHint="Navigates to notifications"
        accessibilityLabel="Activity"
        accessibilityRole="tab"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": false,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            [
              {
                "alignItems": "center",
              },
              {
                "width": "20%",
              },
              {
                "justifyContent": "flex-end",
              },
            ],
          ]
        }
      >
        <View
          style={
            [
              [
                {
                  "height": 44,
                  "width": 44,
                },
                {
                  "alignItems": "center",
                  "justifyContent": "center",
                },
                false,
                undefined,
              ],
            ]
          }
          testID="Notifications"
        >
          <View
            style={
              [
                [
                  {
                    "position": "relative",
                  },
                  {
                    "left": 0.2,
                  },
                  {
                    "top": 0.1,
                  },
                ],
              ]
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#454545",
                    "fontSize": 32,
                  },
                  null,
                  {
                    "fontFamily": "INatIcon",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
        <Text
          maxFontSizeMultiplier={1.2}
          numberOfLines={1}
          style={
            [
              {
                "color": "#454545",
              },
              {
                "fontFamily": "Lato-Regular",
              },
              [
                {
                  "textAlign": "left",
                },
                [
                  {
                    "fontSize": 11,
                    "lineHeight": 13.2,
                  },
                  {
                    "fontFamily": "Lato-Regular",
                  },
                  [
                    {
                      "color": "#454545",
                    },
                  ],
                ],
              ],
            ]
          }
        >
          Activity
        </Text>
      </View>
    </View>
  </View>
</View>
`;
